platform-sdk/docs/proposals/TSS-Library/TSS-Library.md

```diff
@@ -0,0 +1,812 @@
+`TssMessage`: A data structure for distributing encrypted shares of a secret among all participants in a way that only the intended participant can see its part of the share.
+It includes auxiliary information used to validate its correctness and assemble an aggregate public key, i.e., a commitment to a secret share polynomial and a NIZK proof.
+
+###### Generation of the shares
+In the bootstrap process, each participant creates a random EC Private Key `k` out of the Field of the `SignatureScheme` (the secret being shared).
+
+```
+`k`  -  A random EC Private Key for the participant
+`n`  -  The number of total shares across all participants
+`sidᵢ`- The shareId of the share i
+`Xₖ` -  A polynomial with certain properties given a specific secret k
+`sᵢ` -  A point in the polynomial Xₖ for sidᵢ
+```
+
+Then, each shareholder will produce `n` (n=total number of shares) values `Xₛ` by evaluating a polynomial Xₖ at each `ShareId`: `sidᵢ` in the ownership map.
+
+The polynomial `Xₖ` is a polynomial with degree `t-1` (t=threshold) with the form:
+`Xₖ = k + a₁x + ...aₜ₋₁xᵗ⁻¹`[ having: `a₁...aₜ₋₁`: random coefficients from `SignatureScheme.publicKeyGroup` and `k`'s EC field element. x is a field element, thus allowing the polynomial to be evaluated for each share id]
+
+Each `sᵢ = Xₖ(sidᵢ)` constitutes a point on the polynomial.
+
+Once the `sᵢ` value has been calculated for each `ShareId`: `sidᵢ`, the value: `Cᵢ` will be produced by encrypting the `sᵢ` using the `sidᵢ` owner's public key.
+
+![img.svg](img.svg)
+
+The TssMessage will contain all the encrypted values for all shares.
+
+###### Generation of the Polynomial Commitment
+Secret sharing poses a problem for the receiver: did it get a correct share? 
+The dealer may give her a bad share that does not correspond to the dealing, or give so many fake shares to different receiver that they do not correspond to a real dealing. 
+We include a Feldman commitment to the polynomial.
+
+```
+ g = a point generator of `SignatureScheme.publicKeyGroup`
+ aₒ...aₜ₋₁ = coefficients of the polynomial being commited to.
+```
+For each coefficient in the polynomial `Xₖ` `a₍ₒ₎` to `a₍ₜ₋₁₎`, computes a commitment value by calculating: `gᵢ * aᵢ ` (g multiplied by polynomial coefficient `a₍ᵢ₎` )
+
+###### Generation of the NIZKs proofs
+
+Generate a NIZK proof that these commitments and the encrypted shares correspond to a valid sharing of the secret according to the polynomial.
+
+Given the input:
+
+* witness: private data for the proof, including the actual share and the randomness used in the encryption process and private shares(sᵢ).
+
+* statement: The information to proof (sidᵢs,public_keys, polynomial_commitment, encrypted_shares )
+
+* rng: source of randomness
+
+1. Serialize statement
+2. Initialize a pseudorandom number generator using the `SHA256` hash of (1)
+3. `x` = Random out of the random generator in (2)  
+4. Generate random values `α`, `ρ` from the `SignatureScheme.field` out of rng
```

Comment:
Is there an ordering requirement here?

Reply:
In the generation of `α`  and `ρ` values?
I understand there is not.
@rsinha ?
I don't see anything in the code that makes that true. Do you agree?

Reply:
To be specific, I'm asking if all nodes need to use the _same order_, to maintain hash consistency.

Reply:
The statement being serialized and hashed only includes elements computed by the dealer. While there is an ordering amongst those elements as defined by the serialization of that statement, there is no ordering "across" nodes, as the statement is produced by a single node.